### PR TITLE
[SPARK-19104][SQL] Lambda variables in ExternalMapToCatalyst should be global

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

The issue happens in `ExternalMapToCatalyst`. For example, the following codes create `ExternalMapToCatalyst` to convert Scala Map to catalyst map format.

    val data = Seq.tabulate(10)(i => NestedData(1, Map("key" -> InnerData("name", i + 100))))
    val ds = spark.createDataset(data)

The `valueConverter` in `ExternalMapToCatalyst` looks like:

    if (isnull(lambdavariable(ExternalMapToCatalyst_value52, ExternalMapToCatalyst_value_isNull52, ObjectType(class org.apache.spark.sql.InnerData), true))) null else named_struct(name, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(lambdavariable(ExternalMapToCatalyst_value52, ExternalMapToCatalyst_value_isNull52, ObjectType(class org.apache.spark.sql.InnerData), true)).name, true), value, assertnotnull(lambdavariable(ExternalMapToCatalyst_value52, ExternalMapToCatalyst_value_isNull52, ObjectType(class org.apache.spark.sql.InnerData), true)).value)

There is a `CreateNamedStruct` expression (`named_struct`) to create a row of `InnerData.name` and `InnerData.value` that are referred by `ExternalMapToCatalyst_value52`.

Because `ExternalMapToCatalyst_value52` are local variable, when `CreateNamedStruct` splits expressions to individual functions, the local variable can't be accessed anymore.

## How was this patch tested?

Jenkins tests.

### CODE REVIEW

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -736,7 +736,27 @@ class CodegenContext {
       // Cannot split these expressions because they are not created from a row object.
       return expressions.mkString("\n")
     }
-    splitExpressions(expressions, "apply", ("InternalRow", row) :: Nil)
+    splitExpressions(row, expressions, Seq.empty)
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Do we need the check in lines 735-738? Now, we do the same check at lines 754-757, too.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Removed, thanks.

**Comment by viirya** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 2]
```diff
@@ -342,19 +343,38 @@ case class CreateNamedStruct(children: Seq[Expression]) extends CreateNamedStruc
     val values = ctx.freshName("values")
     ctx.addMutableState("Object[]", values, s"$values = null;")
 
+    // `splitExpressions` might split codes to multiple functions. The local variables of
+    // `LambdaVariable` can't be accessed in the functions. We need to add them into the parameters
+    // of the functions.
+    val (valExprCodes, valExprParams) = valExprs.map { expr =>
+      val exprCode = expr.genCode(ctx)
+      val lambdaVars = expr.collect {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: this looks hacky, and why you only do it for `CreateNamedStruct`?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: For example, `CreateArray` is another expression using `splitExpressions`. However, if the external type is an array, `MapObjects` will be used to construct the internal array. `CreateMap` is the same, I think.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: IIRC there are a lot of places we use `splitExpressions`, shall we check all of them?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: I'm not very sure about other places. I move the collecting of lambda variable to `splitExpressions` and let the possible places to do this check.

**Comment by cloud-fan** [comment]
can you put more detailed information about what is the bug? I don't understand why it only happens for map inside case class.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala (hunk 2 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: This approach doesn't work well after rethinking about this and more experiments.

A simpler approach is letting those lambda variables global as MapObjects` and `CollectObjectsToMap` do.

**Review by cloud-fan — APPROVED** [review]
Yea I think this is the right fix according to how we deal with this kind of problem before. LGTM

**Comment by cloud-fan** [comment]
BTW please update PR tile to mention that it's a bug fix for `ExternalMapToCatalyst`

**Comment by viirya** [comment]
@cloud-fan Thanks. The PR title and description are both updated to reflect the change.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetPrimitiveSuite.scala [hunk 3]
```diff
@@ -354,4 +357,9 @@ class DatasetPrimitiveSuite extends QueryTest with SharedSQLContext {
     checkDataset(Seq(PackageClass(1)).toDS(), PackageClass(1))
   }
 
+  test("SPARK-19104: lambda variables should work when parent expression splits generated codes") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetPrimitiveSuite.scala]: you should also update the test name...

Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetPrimitiveSuite.scala]: miss it. updated.

**Comment by viirya** [comment]
retest this please.

**Comment by cloud-fan** [comment]
the test failure is unrelated, merging to master/2.2!

**Pull request closed without merging**